docs/Developers/Design/CRD_design.md

```diff
@@ -0,0 +1,438 @@
+This document describes the set of [Custom Resource Definitions
+(CRDs)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
+that will be used to configure a GCS Gluster cluster. The actual implementation
+of the resources described here will be phased in during development. The
+purpose of this document is to provide the overall structure, ensuring the end
+result provides necessary configurability in a user-friendly manner.
+
+# Overview
+
+A single Gluster operator may control one or more individual Gluster clusters.
+Each cluster could be either hosted within the Kubernetes cluster as a set of
+pods (a converged configuration) or the Gluster nodes could be running on hosts
+outside the Kubernetes cluster (referred to as independent mode). The
+capabilities of the operator differ significantly between these two modes of
+deployment, but the same set of CRDs should be used for both where possible.
+
+A given Gluster cluster is defined by several different Custom Resources (CRs)
+that form a hierarchy. At the top level is the "cluster" CR that describes
+cluster-wide configuration options such as the "name" for the cluster, the TLS
+credentials that will be used for securing communication, and peer clusters for
+geo-replication.
+
+Below the cluster level are zone definitions. These describe the different
+failure domains that the Gluster servers are spread across. This definition
+describes how pods in the zone should be created (for converged deployments) or
+where (DNS name) to find existing nodes (for independent mode).
+
+Below the zone definitions are individual node definitions that track the state
+of the individual Gluster pods. Manipulating these objects permits an
+administrator to place a node into a "quiesced" state for maintenance or
+decommission it entirely.
+
+![Hierarchy of Gluster custom resources](crd_hierarchy.dot.png)
+
+# Custom resources
+
+This section describes the fields in each of the custom resources.
+
+## Cluster CR
+
+The cluster CR defines the cluster-level configuration. A commented example is
+shown below:
+
+```yaml
+apiVersion: "operator.gluster.org/v1alpha1"
+kind: Cluster
+metadata:
+  # Name for the Gluster cluster that will be created by the operator
+  name: my-cluster
+  # CRD is namespaced
+  namespace: gcs
+spec:
+  # Cluster options allows setting "gluster vol set" options that are
+  # cluster-wide (i.e. don't take a volname argument).
+  clusterOptions:  # (optional)
+    "cluster.halo-enabled": "yes"
+  # Gluster CA to use for generating Gluster TLS keys.
+  # Contains Secret w/ CA key & cert
+  glusterCA: ca-secret
+  # Georeplication
+  replication:  # (optional)
+    # Credentials for using this cluster as a target
+    credentials: secretRef
+    targets:
```

Comment:
`target->name` is a cluster name , Isnt it ?  There could be scenarios where one volume is geo replicated between clusters `foo` and `bar` where `foo` as source and `bar` is target. Another volume from `bar` to `foo`. so isnt it a good idea to have a optional `volname` field in targets ?